cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -76,6 +77,22 @@ struct IsInfOperator {
 
 struct IsNullOperator {
   static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    bool is_nan_value = false;
+    bool is_floating = false;
+    if (in.type == float32()) {
```

Comment:
This is comparing pointer values, which is incorrect, even though it will do what you expect most of the time.
(one could e.g. instantiate a separate instance using `std::make_shared<FloatType>()`)